packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.tsx

```diff
@@ -0,0 +1,1057 @@
+                        ? horizontalOptions[0]
+                        : horizontalOptions[1];
+            }
+
+            this.applyBaseHorizontalPositioningState(desiredHorizontalPosition);
+        }
+
+        if (this.props.verticalPositioningMode !== AxisPositioningMode.uncontrolled) {
+            const verticalOptions: ViewportPositionerVerticalPositionLabel[] = this.getVerticalPositioningOptions();
+
+            if (verticalOptions.indexOf(this.state.defaultVerticalPosition) !== -1) {
+                desiredVerticalPosition = this.state.defaultVerticalPosition;
+            }
+
+            const verticalThreshold: number =
+                this.props.verticalThreshold !== undefined
+                    ? this.props.verticalThreshold
+                    : this.positionerRect.height;
+
+            if (
+                desiredVerticalPosition ===
+                    ViewportPositionerVerticalPositionLabel.uncontrolled ||
+                this.getOptionHeight(desiredVerticalPosition) < verticalThreshold
+            ) {
+                desiredVerticalPosition =
+                    this.getOptionHeight(verticalOptions[0]) >
+                    this.getOptionHeight(verticalOptions[1])
+                        ? verticalOptions[0]
+                        : verticalOptions[1];
+            }
+
+            this.applyBaseVerticalPositioningState(desiredVerticalPosition);
+        }
+
+        this.setState({
+            xTranslate: this.calculateHorizontalTranslate(desiredHorizontalPosition),
+            yTranslate: this.calculateVerticalTranslate(desiredVerticalPosition),
+            initialLayoutComplete: true,
+        });
+    };
+
+    /**
+     * Apply base horizontal positioning state based on desired position
+     */
+    private applyBaseHorizontalPositioningState = (
+        desiredHorizontalPosition: ViewportPositionerHorizontalPositionLabel
+    ): void => {
+        let right: number = null;
+        let left: number = null;
+        let xTransformOrigin: string = "left";
+
+        switch (desiredHorizontalPosition) {
+            case ViewportPositionerHorizontalPositionLabel.left:
+                xTransformOrigin = "right";
+                right = this.positionerRect.width;
+                break;
+
+            case ViewportPositionerHorizontalPositionLabel.centerLeft:
+                xTransformOrigin = "right";
+                right = 0;
+                break;
+
+            case ViewportPositionerHorizontalPositionLabel.centerRight:
+                xTransformOrigin = "left";
+                left = 0;
+                break;
+
+            case ViewportPositionerHorizontalPositionLabel.right:
+                xTransformOrigin = "left";
+                left = this.anchorWidth;
+                break;
+        }
+
+        this.setState({
+            xTransformOrigin,
+            right,
+            left,
+            currentHorizontalPosition: desiredHorizontalPosition,
+        });
```

Comment:
you are setting state too in too many locations, this is very hard to track. I would recommend using methods to return new props for the state and setting it a minimum number of times.

Reply:
consolidated some of that.